core/trino-main/src/main/java/io/trino/operator/FlatHash.java

```diff
@@ -153,13 +153,36 @@ public int getCapacity()
         return capacity;
     }
 
+    /**
+     * Releases memory associated with the hash table which is no longer necessary to produce output. Subsequent
+     * calls to insert new elements are rejected, and calls to {@link FlatHash#appendTo(int, BlockBuilder[])} will
+     * incrementally release memory associated with prior groupId values assuming that the caller will only call into
+     * the method to produce output in a sequential fashion.
+     */
+    public void startReleasingOutput()
+    {
+        if (isReleasingOutput()) {
+            throw new IllegalStateException("already releasing output");
+        }
+        control = null;
+        groupIdsByHash = null;
+    }
+
+    private boolean isReleasingOutput()
+    {
+        return control == null;
+    }
+
     public long hashPosition(int groupId)
     {
-        if (groupId < 0) {
-            throw new IllegalArgumentException("groupId is negative");
+        if (groupId < 0 || groupId >= nextGroupId) {
+            throw new IllegalArgumentException("groupId out of range: " + groupId);
         }
         byte[] fixedSizeRecords = getFixedSizeRecords(groupId);
         int fixedRecordOffset = getFixedRecordOffset(groupId);
+        if (isReleasingOutput() && fixedSizeRecords == null) {
```

Comment:
why `fixedSizeRecords == null` check?

use `checkState` instead

Reply:
`fixedSizeRecords` will be null if we've already released the record batch (and are releasing output), otherwise it's valid to refer to the current non-null batch.